analyzers/src/SonarAnalyzer.Common/Rules/Utilities/UtilityAnalyzerBase.cs

```diff
@@ -87,39 +88,48 @@ public abstract class UtilityAnalyzerBase<TSyntaxKind, TMessage> : UtilityAnalyz
         protected UtilityAnalyzerBase(string diagnosticId, string title) : base(diagnosticId, title) { }
 
         protected sealed override void Initialize(SonarAnalysisContext context) =>
-            context.RegisterCompilationAction(c =>
+            context.RegisterCompilationStartAction(startContext =>
+            {
+                ReadParameters(startContext);
+                if (!IsAnalyzerEnabled)
+                {
+                    return;
+                }
+
+                var treeMessages = new List<TMessage>();
+                startContext.RegisterSemanticModelAction(modelContext =>
                 {
-                    ReadParameters(c);
-                    if (!IsAnalyzerEnabled)
+                    if (ShouldGenerateMetrics(modelContext, modelContext.Tree))
                     {
-                        return;
+                        treeMessages.Add(CreateMessage(modelContext.Tree, modelContext.SemanticModel));
                     }
+                });
 
-                    var treeMessages = c.Compilation.SyntaxTrees
-                        .Where(x => ShouldGenerateMetrics(c, x))
-                        .Select(x => CreateMessage(x, c.Compilation.GetSemanticModel(x)));
-                    var messages = CreateAnalysisMessages(c)
+                startContext.RegisterCompilationEndAction(endContext =>
+                {
+                    var allMessages = CreateAnalysisMessages(endContext)
                         .Concat(treeMessages)
                         .WhereNotNull()
                         .ToArray();
                     lock (FileWriteLock)
                     {
                         Directory.CreateDirectory(OutPath);
                         using var stream = File.Create(Path.Combine(OutPath, FileName));
-                        foreach (var message in messages)
+                        foreach (var message in allMessages)
                         {
                             message.WriteDelimitedTo(stream);
                         }
                     }
                 });
+            });
 
         protected virtual bool ShouldGenerateMetrics(SyntaxTree tree) =>
             // The results of Metrics and CopyPasteToken analyzers are not needed for Test projects yet the plugin side expects the protobuf files, so we create empty ones.
             (AnalyzeTestProjects || !IsTestProject)
             && FileExtensionWhitelist.Contains(Path.GetExtension(tree.FilePath))
             && (AnalyzeGeneratedCode || !Language.GeneratedCodeRecognizer.IsGenerated(tree));
 
-        private bool ShouldGenerateMetrics(SonarCompilationReportingContext context, SyntaxTree tree) =>
+        private bool ShouldGenerateMetrics(SonarSematicModelReportingContext context, SyntaxTree tree) =>
```

Comment:
Because `SonarSematicModelReportingContext` has the correct tree, you can use `context.Tree` (I've mentioned it in the previous post)
```suggestion
        private bool ShouldGenerateMetrics(SonarSematicModelReportingContext context) =>
```